### PR TITLE
Initial strict runtime check

### PR DESCRIPTION
**Pull request by naucke**

### CODE REVIEW

File: src/Agda2Hs/Compile/Data.hs [hunk 1]
```diff
@@ -55,17 +61,30 @@ compileData newtyp ds def = do
...
   reportSDoc "agda2hs.data.con" 15 $ text "compileConstructor" <+> prettyTCM c
   reportSDoc "agda2hs.data.con" 20 $ text "  params = " <+> prettyTCM params
   ty <- (`piApplyM` params) . defType =<< getConstInfo c
   reportSDoc "agda2hs.data.con" 20 $ text "  ty = " <+> prettyTCM ty
   TelV tel _ <- telView ty
-  let conName = hsName $ prettyShow $ qnameName c
+
+  let qnameList = qnameToList c
+      -- not the datatype itself
+      qualifier = NE.take (length qnameList - 2) qnameList
```
Comment (naucke) [inline src/Agda2Hs/Compile/Data.hs]: This whole `let` block exists to generate appropriate smart constructor names and I find it a bit ugly, especially the marked line which exists to export the qualified name of the smart constructor _without_ including the name of the data type in the constructor. Is there a better way to do this?

File: src/Agda2Hs/Compile/RuntimeCheckUtils.hs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        )
+        chks
+    successApp = eApp success $ map hsVar args
+    rhs = Hs.GuardedRhss () (Hs.GuardedRhs () [Hs.Qualifier () andChks] successApp : errChks)
+
+-- Creates a run time check if necessary and possible, informing C accordingly.
+-- Takes telescope of dependent type to check, name, and expression on success.
+-- TODO name conflicts
+-- TODO togglable
+checkRtc :: Tele (Dom Type) -> QName -> Hs.Exp () -> C (Maybe (Hs.Decl ()))
```
Comment (naucke) [inline src/Agda2Hs/Compile/RuntimeCheckUtils.hs]: This calls all main functionality. It doesn't handle nested erased types yet.

File: src/Agda2Hs/Compile/RuntimeCheckUtils.hs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    errChks =
+      map
+        ( \chk ->
+            let msg = "Runtime check failed: " ++ Hs.prettyPrint chk
+                hsErr = Hs.App () (hsVar "error") $ Hs.Lit () $ Hs.String () msg msg
+             in Hs.GuardedRhs () [Hs.Qualifier () $ Hs.App () (hsVar "not") chk] hsErr
+        )
+        chks
+    successApp = eApp success $ map hsVar args
+    rhs = Hs.GuardedRhss () (Hs.GuardedRhs () [Hs.Qualifier () andChks] successApp : errChks)
```
Comment (naucke) [inline src/Agda2Hs/Compile/RuntimeCheckUtils.hs]: This is all a bit wordy with the Haskell constructs but I suppose there isn't really a way around it.

File: src/Agda2Hs/Compile/ClassInstance.hs [hunk 4]
```diff
@@ -321,15 +317,6 @@ findDefinitions p m = do
   map snd <$> filterM (uncurry p) inMod
 
 
-resolveStringName :: String -> C QName
```
Comment (naucke) [inline src/Agda2Hs/Compile/ClassInstance.hs:324]: Moved to utils.

File: src/Agda2Hs/Compile/RuntimeCheckUtils.hs [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+-- Create expression to be put in the guard
+createGuardExp :: Type -> C (Maybe (Hs.Exp ()))
+createGuardExp t = do
+  dec <- decify t
+  sth <- findDecInstances dec
+  -- XXX functor something?
+  findDecInstances dec >>= \case
+    Just r -> Just <$> compileTerm dec r
+    Nothing -> return Nothing
```
Comment (naucke) [inline src/Agda2Hs/Compile/RuntimeCheckUtils.hs]: This really feels to me like an fmap should be possible, but I don't know how.

File: src/Agda2Hs/Compile/Name.hs [hunk 6]
```diff
@@ -102,7 +102,8 @@ compileQName f
     parent <- parentName f
     par <- traverse (compileName . qnameName) parent
     let mod0 = qnameModule $ fromMaybe f parent
-    mod <- compileModuleName mod0
+    mod <- compileModuleName mod0 >>= \case
+      Hs.ModuleName () s -> return $ Hs.ModuleName () $ s ++ ".PostRtc"
```
Comment (naucke) [inline src/Agda2Hs/Compile/Name.hs]: This feels a bit error-prone. Is there a better way to do this?

File: src/Agda2Hs/Compile/RuntimeCheckUtils.hs (hunk 5 above)
Comment (liesnikov) [reply src/Agda2Hs/Compile/RuntimeCheckUtils.hs:63]: something like `(fmap $ compileTerm dec r) =<< findDecInstances dec`?

Comment (liesnikov) [reply src/Agda2Hs/Compile/RuntimeCheckUtils.hs]: `copileTerm dec <$> sequence (findDecInstanes dec)`, I think?

Comment (naucke) [reply src/Agda2Hs/Compile/RuntimeCheckUtils.hs]: Nope, `sequence (findDecInstances dec)` expects a `C (C Term)` (idk why exactly), but is a `Maybe (C Term)`

Comment (liesnikov) [reply src/Agda2Hs/Compile/RuntimeCheckUtils.hs]: "expects" as in "has the type"?

Comment (naucke) [reply src/Agda2Hs/Compile/RuntimeCheckUtils.hs]: Just with `sth <- sequence $ findDecInstances dec` (to avoid confusion with the extra error messages that appear with your suggested fmap):
```
    • Couldn't match type ‘Maybe’ with ‘C’
      Expected: C (C Term)
        Actual: Maybe (C Term)
    • In a stmt of a 'do' block: sth <- sequence $ findDecInstances dec
```
(where I replaced
```
                  Control.Monad.Trans.RWS.CPS.RWST
                  Agda2Hs.Compile.Types.CompileEnv
                  Agda2Hs.Compile.Types.CompileOutput
                  Agda2Hs.Compile.Types.CompileState
                  Agda.TypeChecking.Monad.Base.TCM
```
with `C` for brevity)

**Comment by liesnikov** [comment]
No, the error is because you're using it in a C-do block. The compiler is actually suggesting that the term has the right type: `Actual: Maybe (C Term)`
-------- Original message --------From: Jakob Naucke ***@***.***> Date: 19/04/2024  19:19  (GMT+01:00) To: jakobn-ai/agda2hs ***@***.***> Cc: Bohdan Liesnikov ***@***.***>, Comment ***@***.***> Subject: Re: [jakobn-ai/agda2hs] Initial strict runtime check (PR #2) 
@jakobn-ai commented on this pull request.



In src/Agda2Hs/Compile/RuntimeCheckUtils.hs:
> +  findDecInstances dec >>= \case
+    Just r -> Just <$> compileTerm dec r
+    Nothing -> return Nothing

Just with sth <- sequence $ findDecInstances dec (to avoid confusion with the extra error messages that appear with your suggested fmap):
    • Couldn't match type ‘Maybe’ with ‘C’
      Expected: C (C Term)
      Actual: Maybe (C Term)
    • In a stmt of a 'do' block: sth <- sequence $ findDecInstances dec

(where I replaced
                  Control.Monad.Trans.RWS.CPS.RWST
                  Agda2Hs.Compile.Types.CompileEnv
                  Agda2Hs.Compile.Types.CompileOutput
                  Agda2Hs.Compile.Types.CompileState
                  Agda.TypeChecking.Monad.Base.TCM

with C for brevity)

File: src/Agda2Hs/Compile/RuntimeCheckUtils.hs (hunk 5 above)
Comment (naucke) [reply src/Agda2Hs/Compile/RuntimeCheckUtils.hs]: But how can I not be in a `C` block when I need to call `compileTerm :: Type -> Term -> C (Exp ())`?

Comment (liesnikov) [reply src/Agda2Hs/Compile/RuntimeCheckUtils.hs]: the result of `copileTerm dec <$> sequence (findDecInstanes dec)` is a `Maybe (C whatevs)`, if you want to bind it in a `C` do-block I guess you can do another `sequence` on the result?

Alternatively you could try doing
```
r1 <- findDecInstances dec
traverse (compileTerm dec) r1
```

Or maybe `mapMM` ([hackage](https://hackage.haskell.org/package/Agda-2.6.4.3/docs/Agda-Utils-Monad.html#v:mapMM))is what you're after here, to get something like `mapMM (compileTerm dec) (findDecInstances dec)`

Comment (naucke) [reply src/Agda2Hs/Compile/RuntimeCheckUtils.hs]: `traverse`, I was looking for `traverse`. Can even just `findDecInstances dec >>= traverse (compileTerm dec)`. Thank you! Update coming.

Comment (liesnikov) [reply src/Agda2Hs/Compile/RuntimeCheckUtils.hs]: https://impurepics.com/posts/2020-10-03-always-traverse.html

File: src/Agda2Hs/Compile.hs [hunk 7]
```diff
@@ -60,20 +62,22 @@ moduleSetup _ _ m _ = do
 
 compile
   :: Options -> ModuleEnv -> IsMain -> Definition 
-  -> TCM (CompiledDef, CompileOutput)
-compile opts tlm _ def =
+  -> TCM ((CompiledDef, CompiledDef), CompileOutput)
```
Comment (jespercockx) [inline src/Agda2Hs/Compile.hs]: Could you add a comment to explain what the two definitions are?

Comment (jespercockx) [reply src/Agda2Hs/Compile.hs]: Perhaps it would be worth it to define a dedicated record type for pairing the two versions of a definition.

File: src/Agda2Hs/Compile/Data.hs [hunk 8]
```diff
@@ -37,14 +44,23 @@ compileData newtyp ds def = do
...
-    cs <- mapM (compileConstructor params) cs
-    let hd = foldl (Hs.DHApp ()) (Hs.DHead () d) binds
+    chkdCs <- mapM (compileConstructor params) cs
+
+    chks <- ifM (emitsRtc $ defName def) (do
+      let (noneErased, chks) = mapSnd concat $ partitionEithers $ mapMaybe snd chkdCs
+      -- Always export data type name
+      tellNoErased $ prettyName ++ "(" ++ intercalate ", " noneErased ++ ")"
+      return chks)
+      $ return []
```
Comment (jespercockx) [inline src/Agda2Hs/Compile/Data.hs]: use `ifNotM` to avoid trailing `return`s like this.

File: src/Agda2Hs/Compile/Data.hs [hunk 9]
```diff
@@ -37,14 +44,23 @@ compileData newtyp ds def = do
   let params = teleArgs tel
   addContext tel $ do
     binds <- compileTeleBinds tel
-    cs <- mapM (compileConstructor params) cs
-    let hd = foldl (Hs.DHApp ()) (Hs.DHead () d) binds
+    chkdCs <- mapM (compileConstructor params) cs
+
+    chks <- ifM (emitsRtc $ defName def) (do
+      let (noneErased, chks) = mapSnd concat $ partitionEithers $ mapMaybe snd chkdCs
```
Comment (jespercockx) [inline src/Agda2Hs/Compile/Data.hs]: Unclear to me what exactly this code is doing, are you somehow adding the precondition checks from the datatype to the constructors? I don't think that's a useful/necessary thing to do.

File: src/Agda2Hs/Compile/Data.hs [hunk 10]
```diff
@@ -55,17 +71,30 @@ compileData newtyp ds def = do
         DomForall{}     -> genericDocError =<< text "Not supported: indexed datatypes"
       _ -> return ()
 
-compileConstructor :: [Arg Term] -> QName -> C (Hs.QualConDecl ())
+compileConstructor :: [Arg Term] -> QName -> C (Hs.QualConDecl (),
+{- optional exportable cons/checker function -} Maybe (Either String [Hs.Decl ()]))
```
Comment (jespercockx) [inline src/Agda2Hs/Compile/Data.hs]: Try to avoid nested usages of `Maybe` and `Either` like this, usually it's a good sign that you should define a data type yourself.

File: src/Agda2Hs/Compile/Function.hs [hunk 11]
```diff
@@ -162,7 +163,16 @@ compileFun' withSig def@Defn{..} = inTopContext $ withCurrentModule m $ do
         =<< text "Functions defined with absurd patterns exclusively are not supported."
         <+> text "Use function `error` from the Haskell.Prelude instead."
 
-      return $ sig ++ [Hs.FunBind () cs]
+      let def = sig ++ [Hs.FunBind () cs]
+      chk <- ifM (emitsRtc defName) (do
+        typeTel <- theTel <$> telView defType
+        let success = hsVar $ prettyShow m ++ ".PostRtc." ++ prettyShow n
+        checkRtc typeTel defName success >>= \case
+          NoneErased -> do tellNoErased $ prettyShow n; return []
```
Comment (jespercockx) [inline src/Agda2Hs/Compile/Function.hs:168]: This pattern match on the result of `checkRtc` comes back a few times with a very similar structure, perhaps it's worth encapsulating in a helper function.

File: src/Agda2Hs/Compile/Function.hs [hunk 12]
```diff
@@ -376,13 +386,19 @@ withClauseLocals curModule c@Clause{..} k = do
...
+  whenM (emitsRtc (defName def) <&> (&& not (checkNoneErased tele))) $ genericDocError =<<
+        "Cannot make function" <+> prettyTCM (defName def) <+> "transparent." <+>
+        "Transparent functions cannot have erased arguments with runtime checking."
+  compileFun False def >>= (\case
     [Hs.FunBind _ cls] ->
       mapM_ checkTransparentClause cls
     [Hs.TypeDecl _ hd b] ->
       checkTransparentTypeDef hd b
-    _ -> __IMPOSSIBLE__
+    _ -> __IMPOSSIBLE__) . fst
```
Comment (jespercockx) [inline src/Agda2Hs/Compile/Function.hs]: Move `fst` to the `compileFun` call above like this: `(fst <$> compileFun False def) >>= ...` (parenthesis around case expressions are ugly)

File: src/Agda2Hs/Compile/Record.hs [hunk 13]
```diff
@@ -49,7 +54,7 @@ compileMinRecord fieldNames m = do
   -- * it has an explicit dictionary argument
   -- * it's using the fields and definitions from the minimal record and not the parent record
   compiled <- withMinRecord m $ addContext (defaultDom rtype) $ compileLocal $
-    fmap concat $ traverse (compileFun False) defaults
+    concat <$> traverse (fmap (uncurry (++)) . compileFun False) defaults
```
Comment (jespercockx) [inline src/Agda2Hs/Compile/Record.hs]: Why is it fine to just concatenate the pre- and post-rtc functions here?

File: src/Agda2Hs/Compile/Record.hs [hunk 14]
```diff
@@ -100,24 +105,41 @@ compileRecord target def = do
...
+
         when newtyp $ checkNewtypeCon cName fieldDecls
         let target = if newtyp then Hs.NewType () else Hs.DataType ()
-        compileDataRecord constraints fieldDecls target hd ds
-
+        (\c -> ([c], chk)) <$> compileDataRecord constraints fieldDecls target hd ds
   where
-    rName = hsName $ prettyShow $ qnameName $ defName def
-    cName | recNamedCon = hsName $ prettyShow $ qnameName $ conName recConHead
-          | otherwise   = rName   -- Reuse record name for constructor if no given name
```
Comment (jespercockx) [inline src/Agda2Hs/Compile/Record.hs:120]: Please preserve the comment

File: src/Agda2Hs/Compile/RuntimeCheckUtils.hs [hunk 15]
```diff
@@ -0,0 +1,252 @@
...
+    up = "up"
+
+---- Small single-mention convenience functions
+
+-- Gather telescope from domain
+telify :: Dom (a, Type) -> Telescope
+telify = theTel . telView' . snd . unDom
+
+-- Check a type is erased and at top level; in this case, it should be checked.
+checkTopLevelErased :: Dom (a, Type) -> Bool
```
Comment (jespercockx) [inline src/Agda2Hs/Compile/RuntimeCheckUtils.hs]: This is somewhat problematic: since https://github.com/agda/agda2hs/pull/304/ `agda2hs` no longer purely relies on erasure annotations to determine whether to drop an argument.
Ideally, for consistency with the rest of agda2hs we would use the result of `compileDom` to determine whether or not an argument is erased. Otherwise, there is still a risk we miss preconditions in `Prop`, for example

File: src/Agda2Hs/Compile/RuntimeCheckUtils.hs [hunk 16]
```diff
@@ -0,0 +1,252 @@
...
+    >>= \case
+      NoneErased' -> return NoneErased
+      Uncheckable' -> return Uncheckable
+      Checkable' {..} -> do
+        tellAllCheckable name
+        let rhs = eApp success theirRhs
+            chkName = hsName $ prettyShow $ qnameName name
+            chk = createRtc chkName theirLhs theirChks rhs $ binds theirDecls
+        return $ Checkable [chk]
+      . snd
```
Comment (jespercockx) [inline src/Agda2Hs/Compile/RuntimeCheckUtils.hs]: Weird layout, I'm not even sure *what* you are composing the `snd` with

File: src/Agda2Hs/Compile/RuntimeCheckUtils.hs [hunk 17]
```diff
@@ -0,0 +1,252 @@
...
+    -- Use arg name if available, otherwise insert one (consumes one on `argIdx`)
+    (arg, ourArgIdx) = if name == "_" then ("a" ++ show argIdx, succ argIdx) else (name, argIdx)
+    ourLhs = hsPat arg
+    up = "up"
+
+---- Small single-mention convenience functions
+
+-- Gather telescope from domain
+telify :: Dom (a, Type) -> Telescope
+telify = theTel . telView' . snd . unDom
```
Comment (jespercockx) [inline src/Agda2Hs/Compile/RuntimeCheckUtils.hs]: Why is it okay to use `telView'` here instead of `telView`?

File: src/Agda2Hs/Compile/RuntimeCheckUtils.hs [hunk 18]
```diff
@@ -0,0 +1,252 @@
...
+  dec <- resolveStringName "Haskell.Extra.Dec.Dec"
+  level <- liftTCM newLevelMeta
+  let vArg = defaultArg
+      hArg = setHiding Hidden . vArg
+  return $ t {unEl = Def dec $ map Apply [hArg $ Level level, vArg $ unEl t]}
+
+-- Failably find instances for decidable terms
+findDecInstances :: Type -> C (Maybe Term)
+findDecInstances t =
+  liftTCM $
```
Comment (jespercockx) [inline src/Agda2Hs/Compile/RuntimeCheckUtils.hs]: Typically I would put both `liftTCM` and `do` on the same line as `findDecInstances t =`, saving on indentation and vertical space.

File: src/Agda2Hs/Compile/RuntimeCheckUtils.hs [hunk 19]
```diff
@@ -0,0 +1,252 @@
...
+              then return NoneErased'
+              else return Checkable' {..}
+      else return Uncheckable'
+  where
+    doms = telToList tel
+    telsUpTo = map (\i -> fst $ splitTelescopeAt i tel) [0 ..]
+    -- Partition out arguments that are erased and at top level (those we will attempt to check)
+    (topLevelErased, call) = partition (checkTopLevelErased . fst) $ zip doms telsUpTo
+
+-- Check a single type for runtime checkability.
```
Comment (jespercockx) [inline src/Agda2Hs/Compile/RuntimeCheckUtils.hs:248]: Since this is the core logic of this PR, I would appreciate some more detailed commentary with possibly an example to make clear what the `NameIndices` are for. Adding line-by-line explanations to the function definitions would also be a great help, since they are quite difficult to grasp at the moment.

**Review by jespercockx — COMMENTED** [review]
Please find some of my comments on this PR below. I think it's a great addition, however it also adds quite a bit extra complexity to different parts of the `agda2hs` code base. So it would be great to have the opinion of some others (maybe @omelkonian and @ulfnorell) to hear whether they think this is something worth adding (and continue supporting in the future).

File: src/Agda2Hs/Compile/Function.hs (hunk 12 above)
Comment (jakobn-ai) [reply src/Agda2Hs/Compile/Function.hs:398]: `hlint` prefers this, which I agree doesn't make a lot of sense in this instance. I have moved to a separate assignment (also in the case of `RuntimeCheckUtils.checkRtc`)

Comment (naucke) [reply src/Agda2Hs/Compile/Function.hs]: `hlint` prefers this, which I agree doesn't make much sense here. I have moved to a separate assignment (also for `RuntimeCheckUtils.checkRtc`)
(sent single comment earlier by accident, sorry for the noise)

File: src/Agda2Hs/Compile/RuntimeCheckUtils.hs (hunk 16 above)
Comment (naucke) [reply src/Agda2Hs/Compile/RuntimeCheckUtils.hs]: `hlint` prefers this, which I agree doesn't make much sense here. I have moved to a separate assignment (also for `Function.checkTransparentPragma`)

File: src/Agda2Hs/Compile/Record.hs (hunk 13 above)
Comment (naucke) [reply src/Agda2Hs/Compile/Record.hs]: This is only called for classes, which are prohibited. However, it indeed makes more sense to have `fst` (or `defn` in the update with a data type).

File: src/Agda2Hs/Compile/Data.hs (hunk 9 above)
Comment (naucke) [reply src/Agda2Hs/Compile/Data.hs]: Each constructor receives a smart constructor according to its (the constructor's) type, but maybe I don't understand you right here?

Inside the `do` block, the names to be exported are written (always data type plus the constructors with no erased types); all checkable constructors are returned as `chks`.

File: src/Agda2Hs/Compile/RuntimeCheckUtils.hs (hunk 18 above)
Comment (naucke) [reply src/Agda2Hs/Compile/RuntimeCheckUtils.hs]: Hmmm, I really enjoyed working with autoformatters and all the major ones seem to do this. I've moved the `liftTCM` to the caller, but let me know if you find that unclear.

File: src/Agda2Hs/Compile/Function.hs (hunk 11 above)
Comment (naucke) [reply src/Agda2Hs/Compile/Function.hs:168]: Hmm, there are subtle differences though which is why I designed the signatures in this way:

|           | None erased                 | All checkable               |
|-----------|-----------------------------|-----------------------------|
| Function  | mark for export             | prepend signature (list)    |
| Data type | mark constructor for export | prepend signature (element) |
| Record    | mark constructor for export | have no signature           |

File: src/Agda2Hs/Compile/RuntimeCheckUtils.hs (hunk 19 above)
Comment (naucke) [reply src/Agda2Hs/Compile/RuntimeCheckUtils.hs:248]: I have enhanced the comments, do let me know if you feel this is sufficiently illustrated.

File: src/Agda2Hs/Compile/RuntimeCheckUtils.hs (hunk 17 above)
Comment (naucke) [reply src/Agda2Hs/Compile/RuntimeCheckUtils.hs]: (as discussed in person) it should be fine to always have `telView'` because we don't need type reduction. I have updated the rest of the PR accordingly.
